ligo/main.mligo

```diff
@@ -140,15 +139,76 @@ let calc_fee_growth_inside (s : storage) (lower_tick_index : tick_index) (upper_
     let position = {position with fee_growth_inside_last = fee_growth_inside} in
-    let positions = Big_map.update key (Some position) s.positions in
+    let positions = Big_map.add key position s.positions in
     ({s with positions = positions}, fees, position)
 
+
+let update_balances_after_position_change
+        (s : storage)
+        (lower_tick_index : tick_index) (upper_tick_index : tick_index)
+        (maximum_tokens_contributed : balance_nat)
+        (to_x : address) (to_y : address)
+        (liquidity_delta : int) (fees : balance_nat) : result =
+    (* Compute how much should be deposited / withdrawn to change liquidity by liquidity_net *)
+
+    (* Grab cached prices for the interval *)
+    let ticks = s.ticks in
+    let tick_u = get_tick ticks upper_tick_index internal_tick_not_exist_err in
+    let tick_l = get_tick ticks lower_tick_index internal_tick_not_exist_err in
+    let srp_u = tick_u.sqrt_price in
+    let srp_l = tick_l.sqrt_price in
+
+    (* Add or remove liquidity above the current tick *)
+    let (s, delta) =
+    if s.cur_tick_index.i < lower_tick_index.i then
+        (s, {
+            (* If I'm adding liquidity, x will be positive, I want to overestimate it, if x I'm taking away
+                liquidity, I want to to underestimate what I'm receiving. *)
+            x = ceildiv_int (liquidity_delta * (int (Bitwise.shift_left (assert_nat (srp_u.x80 - srp_l.x80, internal_sqrt_price_grow_err_1)) 80n))) (int (srp_l.x80 * srp_u.x80)) ;
+            y = 0})
+    else if lower_tick_index.i <= s.cur_tick_index.i && s.cur_tick_index.i < upper_tick_index.i then
+        (* update interval we are in, if need be ... *)
+        let s = { s with
+                    liquidity = assert_nat (s.liquidity + liquidity_delta, position_liquidity_below_zero_err)
+                } in
+        (s, {
+            x = ceildiv_int (liquidity_delta * (int (Bitwise.shift_left (assert_nat (srp_u.x80 - s.sqrt_price.x80, internal_sqrt_price_grow_err_2)) 80n))) (int (s.sqrt_price.x80 * srp_u.x80)) ;
+            y = shift_int (liquidity_delta * (s.sqrt_price.x80 - srp_l.x80)) (-80)
+            })
+    else (* cur_tick_index >= p.upper_tick_index *)
+        (s, {x = 0 ; y = shift_int (liquidity_delta * (srp_u.x80 - srp_l.x80)) (-80) }) in
+
+    (* Collect fees to increase withdrawal or reduce required deposit. *)
+    let delta = {x = delta.x - fees.x ; y = delta.y - fees.y} in
+
+    (* Check delta doesn't exceed maximum_tokens_contributed. *)
+    let _: unit = if delta.x > int(maximum_tokens_contributed.x) then failwith high_tokens_err else unit in
+    let _: unit = if delta.y > int(maximum_tokens_contributed.y) then failwith high_tokens_err else unit in
+
+    let op_x = if delta.x > 0 then
+        x_transfer Tezos.sender Tezos.self_address (abs delta.x) s.constants
+    else
+#if DEBUG
+        let _ : unit = if to_x = Tezos.self_address then failwith internal_unexpected_income_err else unit in
+#endif
```

Comment:
This seems a pretty good idea in general as well.

We may abstract the pattern of `failwith`/`abs` for internal errors that result from conversion from `int`s to `nat`s :thinking: